### ISSUE TITLE
git: protocol for submodules may fail if ports are blocked

### ISSUE THREAD

**Issue by mouse07410** [objectcomputing/OpenDDS#1481]
On MacOS I found a weird problem. In theory, `git://github.com/blahblahblah.git` should work. However, apparently it doesn't. Also, IntelliJ IDEA IDE refuses URLs in such form.

To remedy, the URL should be changed to either pulling via HTTPS, aka `https://github.com/blahblahblah.git`, or if SSH pull is preferred - to this form: `ssh://git@github.com/blahblahblah.git`.

Here's the diff to change it to HTTPS (it still works on other platforms, but now will also work on MacOS):
```diff
--- .gitmodules.orig	2020-01-25 17:00:22.000000000 -0500
+++ .gitmodules	2020-01-25 17:00:42.000000000 -0500
@@ -1,7 +1,7 @@
 [submodule "tools/rapidjson"]
 	path = tools/rapidjson
-	url = git://github.com/miloyip/rapidjson.git
+	url = https://github.com/miloyip/rapidjson.git
 	branch = version1.1.0
 [submodule "tests/googletest"]
 	path = tests/googletest
-	url = git://github.com/google/googletest
+	url = https://github.com/google/googletest
```

**Comment by iguessthislldo** [comment]
I don't think changing the protocol would hurt, but this sounds like an issue with your git client. My CLI git version 2.17 has no problems with the OpenDDS submodules. Also I would think a git client would support the git protocol.

**Comment by mitza-oci** [comment]
No problems here:
```
$ git --version
git version 2.21.1 (Apple Git-122.3)
$ sw_vers -productVersion
10.15.2
```
I assume this is network related.

**Comment by mouse07410** [comment]
> I assume this is network related.

I strongly doubt that, as there hasn't been any supporting evidence. Also:
```
$ git --version
git version 2.25.0
$ sw_vers -productVersion
10.14.6
```

> I don't think changing the protocol would hurt, but this sounds like an issue with your git client.

It might be the client issue - but since, as we all seem to agree, it wouldn't hurt to change the protocol, let's just do that, please?

**Comment by iguessthislldo** [comment]
Can you tell us exactly what your git client(s) are saying?

**Comment by mouse07410** [comment]
You bet:
```
$ git clone https://github.com/objectcomputing/OpenDDS.git ODDS-test
Cloning into 'ODDS-test'...
remote: Enumerating objects: 152, done.
remote: Counting objects: 100% (152/152), done.
remote: Compressing objects: 100% (133/133), done.
remote: Total 169292 (delta 81), reused 46 (delta 19), pack-reused 169140
Receiving objects: 100% (169292/169292), 67.73 MiB | 1.29 MiB/s, done.
Resolving deltas: 100% (131264/131264), done.
Updating files: 100% (6370/6370), done.
$ cd ODDS-test
$ git submodule update --init --recursive
Submodule 'tests/googletest' (git://github.com/google/googletest) registered for path 'tests/googletest'
Submodule 'tools/rapidjson' (git://github.com/miloyip/rapidjson.git) registered for path 'tools/rapidjson'
Cloning into '/Users/ur20980/src/ODDS-test/tests/googletest'...
fatal: read error: Connection reset by peer
fatal: clone of 'git://github.com/google/googletest' into submodule path '/Users/ur20980/src/ODDS-test/tests/googletest' failed
Failed to clone 'tests/googletest'. Retry scheduled
Cloning into '/Users/ur20980/src/ODDS-test/tools/rapidjson'...
fatal: read error: Connection reset by peer
fatal: clone of 'git://github.com/miloyip/rapidjson.git' into submodule path '/Users/ur20980/src/ODDS-test/tools/rapidjson' failed
Failed to clone 'tools/rapidjson'. Retry scheduled
Cloning into '/Users/ur20980/src/ODDS-test/tests/googletest'...
fatal: read error: Connection reset by peer
fatal: clone of 'git://github.com/google/googletest' into submodule path '/Users/ur20980/src/ODDS-test/tests/googletest' failed
Failed to clone 'tests/googletest' a second time, aborting
$ 
```

And after replacing `git://...` with `ssh://git@...`:
```
$ git submodule update --init --recursive
Cloning into '/Users/ur20980/src/ODDS-test/tools/rapidjson'...
Enter PIN for '[my smartcard with SSH keys]': 
Submodule path 'tests/googletest': checked out 'dea0216d0c6bc5e63cf5f6c8651cd268668032ec'
Submodule path 'tools/rapidjson': checked out '2fb78f9cee01996f020d5af7592b37c5a0693b31'
Submodule 'thirdparty/gtest' (https://github.com/google/googletest.git) registered for path 'tools/rapidjson/thirdparty/gtest'
Cloning into '/Users/ur20980/src/ODDS-test/tools/rapidjson/thirdparty/gtest'...
Submodule path 'tools/rapidjson/thirdparty/gtest': checked out '0a439623f75c029912728d80cb7f1b8b48739ca4'
$
```

**Comment by mouse07410** [comment]
Is there a problem applying [this](https://github.com/objectcomputing/OpenDDS/issues/1481#issue-555144449) patch?

**Comment by mitza-oci** [comment]
> 
> 
> Is there a problem applying [this](https://github.com/objectcomputing/OpenDDS/issues/1481#issue-555144449) patch?

Well we know it's not a macOS issue.  In theory this would be platform-independent and would depend on your host being able to reach `github.com`'s server using the git:// protocol.

[Git's documentation](https://git-scm.com/book/en/v2/Git-on-the-Server-The-Protocols#The-Git-Protocol) states that `git://` is "the fastest network transfer protocol available".  Especially since these submodules are used on time-limited CI servers, that's of some value.

This idea https://www.gniibe.org/memo/software/git/using-submodule.html might be good to try out, if it works in this case.

**Comment by simpsont-oci** [comment]
Have you tried running 'git gc' on your repo and then trying again with the
git:// URL? That seemed to fix the "fatal: read error: Connection reset by
peer" issue for some other folks on stackoverflow
<https://stackoverflow.com/questions/17977206/fatal-read-error-connection-reset-by-peer>
.

**Comment by mouse07410** [comment]
> Have you tried running 'git gc' on your repo and then trying again with the
git:// URL? That seemed to fix the "fatal: read error: Connection reset by
peer" issue for some other folks on stackoverflow...

I haven't tried that. In general, it doesn't seem reasonable that a freshly-cloned repo would require `git gc`. 
But I also noticed that the very next comment in that thread mentioned the fix I came up with. 

Now, to the root of the probem with `git://`, as [Git's documentation](https://git-scm.com/book/en/v2/Git-on-the-Server-The-Protocols#The-Git-Protocol) describes:
> ...It also requires firewall access to port 9418, which isn’t a standard port that corporate firewalls always allow. Behind big corporate firewalls, this obscure port is commonly blocked...

I'm doing this from behind a big corporate firewall - and no, that port isn't likely to get opened any time soon, even if I ask for it. ;-)

Again, I request applying the above patch.
